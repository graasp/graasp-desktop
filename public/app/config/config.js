// eslint-disable-next-line import/no-extraneous-dependencies
const { app } = require('electron');
const isWindows = require('../utils/isWindows');

// types that we support downloading
const DOWNLOADABLE_MIME_TYPES = [
  // video
  'application/mp4',
  'application/ogg',
  'video/mp4',
  'video/ogg',
  'video/quicktime',
  'video/webm',
  // audio
  'audio/mp4',
  'audio/mpeg',
  'audio/ogg',
  'audio/webm',
  'audio/x-aac',
  'audio/x-wav',
  // image
  'image/bmp',
  'image/gif',
  'image/jpeg',
  'image/png',
  // pdf
  'application/pdf',
];

// resolve path for windows '\'
const escapeEscapeCharacter = str => {
  return isWindows() ? str.replace(/\\/g, '\\\\') : str;
};

// categories
const RESOURCE = 'Resource';
const APPLICATION = 'Application';

const VAR_FOLDER = `${escapeEscapeCharacter(app.getPath('userData'))}/var`;
const DATABASE_PATH = `${VAR_FOLDER}/db.json`;
const ICON_PATH = 'app/assets/icon.png';
const PRODUCT_NAME = 'Graasp';
const TMP_FOLDER = 'tmp';
const DEFAULT_LANG = 'en';
const DEFAULT_DEVELOPER_MODE = false;
const DEFAULT_GEOLOCATION_ENABLED = false;
const DEFAULT_PROTOCOL = 'https';

module.exports = {
  DEFAULT_PROTOCOL,
  DEFAULT_DEVELOPER_MODE,
  DEFAULT_GEOLOCATION_ENABLED,
  DOWNLOADABLE_MIME_TYPES,
  TMP_FOLDER,
  RESOURCE,
  APPLICATION,
  DATABASE_PATH,
  VAR_FOLDER,
  DEFAULT_LANG,
  ICON_PATH,
  PRODUCT_NAME,
  escapeEscapeCharacter,
};
