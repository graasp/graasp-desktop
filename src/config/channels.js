// this file needs to use module.exports as it is used both by react and
// electron make sure this file is identical in both src/config/channels.js
// and public/app/config/channels.js

module.exports = {
  SAVE_SPACE_CHANNEL: 'space:save',
  GET_SPACE_CHANNEL: 'space:get',
  GET_SPACES_CHANNEL: 'spaces:get',
  DELETE_SPACE_CHANNEL: 'space:delete',
  DELETED_SPACE_CHANNEL: 'space:deleted',
  LOAD_SPACE_CHANNEL: 'space:load',
  LOADED_SPACE_CHANNEL: 'space:loaded',
  EXPORT_SPACE_CHANNEL: 'space:export',
  EXPORTED_SPACE_CHANNEL: 'space:exported',
  SHOW_LOAD_SPACE_PROMPT_CHANNEL: 'prompt:space:load:show',
  SHOW_EXPORT_SPACE_PROMPT_CHANNEL: 'prompt:space:export:show',
  RESPOND_LOAD_SPACE_PROMPT_CHANNEL: 'prompt:space:load:response',
  RESPOND_EXPORT_SPACE_PROMPT_CHANNEL: 'prompt:space:export:respond',
  SHOW_DELETE_SPACE_PROMPT_CHANNEL: 'prompt:space:delete:show',
  RESPOND_DELETE_SPACE_PROMPT_CHANNEL: 'prompt:space:delete:respond',
  GET_USER_FOLDER_CHANNEL: 'user:folder:get',
  GET_LANGUAGE_CHANNEL: 'user:lang:get',
  SET_LANGUAGE_CHANNEL: 'user:lang:set',
  SET_DEVELOPER_MODE_CHANNEL: 'user:developer-mode:set',
  GET_DEVELOPER_MODE_CHANNEL: 'user:developer-mode:get',
  GET_APP_INSTANCE_RESOURCES_CHANNEL: 'app-instance-resources:get',
  POST_APP_INSTANCE_RESOURCE_CHANNEL: 'app-instance-resource:post',
  PATCH_APP_INSTANCE_RESOURCE_CHANNEL: 'app-instance-resource:patch',
  DELETE_APP_INSTANCE_RESOURCE_CHANNEL: 'app-instance-resource:delete',
  GET_APP_INSTANCE_CHANNEL: 'app-instance:get',
  PATCH_APP_INSTANCE_CHANNEL: 'app-instance:patch',
  GET_DATABASE_CHANNEL: 'developer:database:get',
  SET_DATABASE_CHANNEL: 'developer:database:set',
  SHOW_SYNC_SPACE_PROMPT_CHANNEL: 'prompt:space:sync:show',
  RESPOND_SYNC_SPACE_PROMPT_CHANNEL: 'prompt:space:sync:respond',
  SYNC_SPACE_CHANNEL: 'space:sync',
  SYNCED_SPACE_CHANNEL: 'space:synced',
};
