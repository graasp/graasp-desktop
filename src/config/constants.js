// phase item types
export const TEXT = 'text/html';
export const VIDEO = new RegExp('video/*');
export const IMAGE = new RegExp('image/*');
export const RESOURCE = 'Resource';
export const APPLICATION = 'Application';
export const IFRAME = 'application/octet-stream';
export const DEFAULT_RADIUS = 50;
export const DEFAULT_LANGUAGE = 'en';
export const DEFAULT_DEVELOPER_MODE = false;
export const DEFAULT_GEOLOCATION_ENABLED = false;
export const SHORT_ID_LENGTH = 6;
export const SMART_GATEWAY_HOST = 'gateway.golabz.eu';
export const SMART_GATEWAY_QUERY_STRING_DIVIDER = '#';
export const CONTROL_TYPES = {
  BUTTON: 'BUTTON',
  SWITCH: 'SWITCH',
};
export const MIN_CARD_WIDTH = 345;
