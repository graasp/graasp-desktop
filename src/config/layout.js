export const MAX_APP_HEIGHT = 900;
export const MIN_APP_HEIGHT = 50;
export const DEFAULT_APP_HEIGHT = 300;
export const DEFAULT_TOOLS_WIDTH = 600;
export const MAX_TOOLS_WIDTH = 1200;
export const MIN_TOOLS_WIDTH = 100;
