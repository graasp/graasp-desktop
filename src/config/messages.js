// this file needs to use module.exports as it is used both by react and
// electron make sure this file is identical in both src/config/messages.js
// and public/app/config/messages.js

const ERROR_DOWNLOADING_MESSAGE = 'There was a problem downloading your files';
const ERROR_LOADING_MESSAGE = 'There was a problem loading your space';
const ERROR_EXPORTING_MESSAGE = 'There was a problem exporting this space';
const SUCCESS_EXPORTING_MESSAGE = 'Space was exported successfully';
const ERROR_DELETING_MESSAGE = 'There was a problem deleting this space';
const SUCCESS_DELETING_MESSAGE = 'Space was deleted successfully';
const ERROR_ZIP_CORRUPTED_MESSAGE =
  'The archive provided is not formatted properly';
const ERROR_JSON_CORRUPTED_MESSAGE = "Space's JSON file is corrupted";
const ERROR_SPACE_ALREADY_AVAILABLE_MESSAGE =
  'A space with the same id is already available';
const SUCCESS_SPACE_LOADED_MESSAGE = 'Space was loaded successfully';
const UNEXPECTED_ERROR_MESSAGE = 'An unexpected error has occurred';
const ERROR_GETTING_SPACE_MESSAGE = 'There was an error getting that space.';
const OFFLINE_ERROR_MESSAGE = 'This functionality requires online connectivity';
const ERROR_MESSAGE_HEADER = 'Error';
const ERROR_SAVING_SPACE_MESSAGE = 'There was an error saving the space';
const SUCCESS_SAVING_MESSAGE = 'Space was saved successfully';
const SUCCESS_MESSAGE_HEADER = 'Success';
const ERROR_GETTING_GEOLOCATION =
  'There was an error getting your current location';
const ERROR_GETTING_SPACES_NEARBY =
  'There was an error getting the spaces nearby';
const ERROR_GETTING_USER_FOLDER =
  'There was an error getting your user folder.';
const ERROR_GETTING_LANGUAGE = 'There was an error getting the language.';
const ERROR_SETTING_LANGUAGE = 'There was an error setting the language.';
const INVALID_SPACE_ID = 'Invalid space ID.';
const ERROR_GETTING_DEVELOPER_MODE =
  'There was an error getting the developer mode';
const ERROR_SETTING_DEVELOPER_MODE =
  'There was an error setting the developer mode';
const ERROR_GETTING_DATABASE = 'There was an error getting the database.';
const ERROR_SETTING_DATABASE = 'There was an error updating the database.';
const SUCCESS_SYNCING_MESSAGE = 'Space was successfully synced';
const ERROR_SYNCING_MESSAGE = 'There was an error syncing the space.';

module.exports = {
  ERROR_GETTING_DEVELOPER_MODE,
  ERROR_SETTING_DEVELOPER_MODE,
  ERROR_GETTING_LANGUAGE,
  ERROR_SETTING_LANGUAGE,
  ERROR_DOWNLOADING_MESSAGE,
  ERROR_MESSAGE_HEADER,
  ERROR_SAVING_SPACE_MESSAGE,
  SUCCESS_SAVING_MESSAGE,
  SUCCESS_MESSAGE_HEADER,
  ERROR_GETTING_SPACE_MESSAGE,
  ERROR_EXPORTING_MESSAGE,
  SUCCESS_EXPORTING_MESSAGE,
  ERROR_DELETING_MESSAGE,
  SUCCESS_DELETING_MESSAGE,
  ERROR_ZIP_CORRUPTED_MESSAGE,
  ERROR_JSON_CORRUPTED_MESSAGE,
  ERROR_SPACE_ALREADY_AVAILABLE_MESSAGE,
  SUCCESS_SPACE_LOADED_MESSAGE,
  UNEXPECTED_ERROR_MESSAGE,
  OFFLINE_ERROR_MESSAGE,
  ERROR_LOADING_MESSAGE,
  ERROR_GETTING_GEOLOCATION,
  ERROR_GETTING_SPACES_NEARBY,
  ERROR_GETTING_USER_FOLDER,
  INVALID_SPACE_ID,
  ERROR_GETTING_DATABASE,
  ERROR_SETTING_DATABASE,
  SUCCESS_SYNCING_MESSAGE,
  ERROR_SYNCING_MESSAGE,
};
