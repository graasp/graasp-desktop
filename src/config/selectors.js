export const ROOT_ID = 'root';

export const DRAWER_BUTTON_ID = 'drawerButton';

export const FAVORITE_SPACES_WRAPPER_ID = 'favoriteSpacesWrapper';
export const RECENT_SPACES_WRAPPER_ID = 'recentSpacesWrapper';

export const MAIN_MENU_ID = 'mainMenu';
export const CLASSROOMS_MENU_ITEM_ID = 'classroomsMenuItem';
export const CLASSROOMS_MAIN_ID = 'classroomsMain';
export const SAVED_SPACES_MENU_ITEM_ID = 'savedSpacesMenuItem';
export const SPACES_NEARBY_MENU_ITEM_ID = 'spacesNearbyMenuItem';
export const SPACES_NEARBY_MAIN_ID = 'spacesNearbyMain';
export const SAVED_SPACES_MAIN_ID = 'savedSpacesMain';
export const VISIT_MENU_ITEM_ID = 'visitMenuItem';
export const VISIT_SPACE_BUTTON_CLASS = 'VisitSpaceButton';
export const VISIT_SPACE_INPUT_CLASS = 'VisitSpaceInput';
export const VISIT_MAIN_ID = 'visit';
export const LOAD_MENU_ITEM_ID = 'loadMenuItem';
export const LOAD_MAIN_ID = 'load';
export const SETTINGS_MENU_ITEM_ID = 'settingsMenuItem';
export const SETTINGS_MAIN_ID = 'settings';
export const SETTINGS_ACTIONS_CLASS = 'SettingsActions';
export const VISIT_INPUT_ID = 'visitInput';
export const VISIT_BUTTON_ID = 'visitButton';
export const SPACE_TOOLBAR_ID = 'spaceToolbar';
export const HOME_MENU_ITEM_ID = 'homeMenuItem';
export const HOME_MAIN_ID = 'homeMain';
export const DASHBOARD_MENU_ITEM_ID = 'dashboardMenuItem';
export const DASHBOARD_MAIN_ID = 'dashboardMain';
export const DEVELOPER_MENU_ITEM_ID = 'developerMenuItem';
export const DEVELOPER_MAIN_ID = 'developerMain';
export const SIGN_OUT_MENU_ITEM_ID = 'signOutMenuItem';
export const SIGN_IN_MAIN_ID = 'signInMain';

export const PHASE_MENU_LIST_ID = 'phaseMenuList';
export const PHASE_MENU_ITEM_HOME_ID = 'phaseMenuItemHome';

export const SPACE_SCREEN_CLASS = 'SpaceScreen';
export const SPACE_THUMBNAIL_IMAGE_CLASS = 'spaceThumbnailImageClass';
export const SPACE_START_PREVIEW_BUTTON = 'spaceStartPreviewButton';
export const SPACE_DESCRIPTION_ID = 'spaceDescription';
export const SPACE_SAVE_ICON_CLASS = 'spaceSaveIcon';
export const SPACE_PREVIEW_ICON_CLASS = 'spacePreviewIcon';
export const PHASE_DESCRIPTION_ID = 'phaseDescription';
export const PHASE_DRAWER_CLASS = 'PhasesDrawer';
export const SPACE_MEDIA_CARD_CLASS = 'spaceMediaCard';
export const buildPhaseMenuItemId = (id) => `phaseMenuItemId-${id}`;
export const BANNER_WARNING_PREVIEW_ID = 'bannerWarningPreview';
export const SPACE_EXPORT_BUTTON_CLASS = 'spaceExportButton';
export const SPACE_DELETE_BUTTON_CLASS = 'spaceDeleteButton';
export const SPACE_CLEAR_BUTTON_CLASS = 'spaceClearButton';
export const SPACE_SYNC_BUTTON_CLASS = 'spaceSyncButton';
export const SPACE_FAVORITE_BUTTON_CLASS = 'spaceFavoriteButton';
export const TOOLS_BUTTON_CLASS = 'toolsButton';
export const TOOLS_CONTENT_PANE_ID = 'toolsContentPane';
export const SPACE_SEARCH_INPUT_ID = 'searchInput';

export const SPACE_CARD_CLASS = 'spaceCard';
export const buildSpaceCardId = (id) => `spaceCard-${id}`;
export const SPACE_DESCRIPTION_EXPAND_BUTTON_CLASS =
  'spaceDescriptionExpandButton';
export const buildSpaceCardDescriptionId = (id) => `spaceCardDescription-${id}`;
export const SPACE_CARD_LINK_CLASS = 'spaceCardLink';
export const SPACE_NOT_AVAILABLE_TEXT_ID = 'noSpaceAvailableText';

export const LOAD_BROWSE_BUTTON_ID = 'loadBrowseButton';
export const LOAD_INPUT_ID = 'loadInput';
export const LOAD_SUBMIT_BUTTON_ID = 'loadSubmitButton';
export const LOAD_LOAD_BUTTON_ID = 'loadLoadButton';
export const LOAD_BACK_BUTTON_ID = 'loadCancelButton';
export const buildCheckboxLabel = (name) => `checkboxLabel-${name}`;

export const EXPORT_SPACE_BUTTON_ID = 'exportSpaceButton';
export const EXPORT_SPACE_BACK_BUTTON_ID = 'exportSpaceBackButton';

export const SETTINGS_TITLE_ID = 'settingsTitle';
export const LANGUAGE_SELECT_ID = 'settingsLanguageSelect';
export const DEVELOPER_SWITCH_ID = 'settingsDeveloperSwitch';
export const GEOLOCATION_CONTROL_ID = 'settingsGeolocationControl'; // todo: remove
export const SYNC_MODE_SWITCH_ID = 'syncModeSwitch';
export const STUDENT_MODE_SWITCH_ID = 'studentModeSwitch';

export const buildPhaseAppName = (id) => `phaseApp-${id}`;

export const LOGIN_USERNAME_INPUT_ID = 'loginUsernameInput';
export const LOGIN_BUTTON_ID = 'loginButton';

export const SYNC_ITEM_CLASS = 'syncItem';
export const SPACE_DESCRIPTION_TEXT_CLASS = 'SpaceDescriptionText';
export const SYNC_ACCEPT_BUTTON_ID = 'syncAcceptButton';
export const SYNC_CANCEL_BUTTON_ID = 'syncCancelButton';
export const SYNC_ADVANCED_MAIN_ID = 'syncAdvancedMain';
export const SYNC_VISUAL_MAIN_ID = 'syncVisualMain';

export const NO_CLASSROOM_AVAILABLE_ID = 'noClassroomAvailable';
export const ADD_CLASSROOM_BUTTON_ID = 'addClassroomButton';
export const CLASSROOM_NAME_INPUT_ID = 'classroomNameInput';
export const ADD_CLASSROOM_VALIDATE_BUTTON_ID = 'addClassroomValidateButton';
export const ADD_CLASSROOM_CANCEL_BUTTON_ID = 'addClassroomCancelButton';
export const CLASSROOM_CARD_CLASS = 'classroomCard';
export const buildClassroomCardId = (id) => `classroomCard-${id}`;
export const EDIT_CLASSROOM_BUTTON_CLASS = 'editClassroomButton';
export const EDIT_CLASSROOM_INPUT_ID = 'editClassroomInput';
export const EDIT_CLASSROOM_VALIDATE_BUTTON_ID = 'editClassroomValidateButton';
export const EDIT_CLASSROOM_CANCEL_BUTTON_ID = 'editClassroomCancelButton';
export const DELETE_CLASSROOM_BUTTON_CLASS = 'deleteClassroomButton';
export const EDIT_USER_IN_CLASSROOM_DIALOG_TITLE_ID =
  'editUserInClassroomDialogTitle';
export const ADD_USER_IN_CLASSROOM_DIALOG_TITLE_ID =
  'addUserInClassroomDialogTitle';
export const ADD_USER_IN_CLASSROOM_BUTTON_ID = 'addUserInClassroomButton';
export const ADD_USER_IN_CLASSROOM_NAME_INPUT_ID =
  'addUserInClassroomNameInput';
export const ADD_USER_IN_CLASSROOM_VALIDATE_BUTTON_ID =
  'addUserInClassroomValideButton';
export const ADD_USER_IN_CLASSROOM_CANCEL_BUTTON_ID =
  'addUserInClassroomCancelButton';
export const EDIT_USER_IN_CLASSROOM_BUTTON_CLASS = 'editUserInClassroomButton';
export const EDIT_USER_IN_CLASSROOM_USERNAME_INPUT_ID =
  'editUserInClassroomUsernameInput';
export const EDIT_USER_IN_CLASSROOM_VALIDATE_BUTTON_ID =
  'editUserInClassroomValidateButton';
export const CLASSROOM_TABLE_BODY_ID = 'classroomTableBody';
export const DELETE_USER_IN_CLASSROOM_BUTTON_CLASS =
  'deleteUserInClassroomButton';
export const CLASSROOM_CARD_SPACES_CLASS = 'classroomCardSpaces';
export const CLASSROOM_CARD_STUDENTS_CLASS = 'classroomCardStudents';
export const CLASSROOM_SCREEN_BACK_BUTTON_ID = 'classroomScreenBackButton';
export const CLASSROOM_CARD_NAME_CLASS = 'classroomCardName';
export const SELECT_USER_IN_CLASSROOM_CLASS = 'selectUserInClassroom';
export const DELETE_USERS_IN_CLASSROOM_BUTTON_ID =
  'deleteUsersInClassroomButton';
export const IMPORT_DATA_IN_CLASSROOM_BUTTON_ID = 'importDataInClassroomButton';
export const IMPORT_FILEPATH_IN_CLASSROOM_INPUT_ID =
  'importFilepathInClassroomInput';
export const IMPORT_DATA_IN_CLASSROOM_SUBMIT_BUTTON_ID =
  'importDataInClassroomSubmitButton';
export const IMPORT_DATA_IN_CLASSROOM_STUDENT_SELECT_CONTAINER_CLASS =
  'importDataInClassroomStudentSelectContainer';
export const IMPORT_DATA_IN_CLASSROOM_STUDENT_SELECT_PREFIX_CLASS =
  'importDataInClassroomStudentSelectPrefix';
export const IMPORT_DATA_IN_CLASSROOM_BACK_BUTTON_ID =
  'importDataInClassroomBackButton';
export const IMPORT_DATA_CLASSROOM_VALIDATE_BUTTON_ID =
  'importDataInClassroomValidateButton';
export const STUDENT_ROW_RESOURCES_CLASS = 'studentRowResources';
export const STUDENT_ROW_ACTIONS_CLASS = 'studentRowActions';
export const EDIT_USER_IN_CLASSROOM_DELETE_DATA_BUTTON_CLASS =
  'editUserInClassroomDeleteDataButton';
export const EDIT_CLASSROOM_DELETE_DATA_BUTTON_CLASS =
  'editClassroomDeleteDataButton';
export const buildTableCellSpaceId = (id) => `tableCellSpace-${id}`;

export const DRAWER_HEADER_TEACHER_ICON_ID = 'drawerHeaderTeacherIcon';
export const DRAWER_HEADER_STUDENT_ICON_ID = 'drawerHeaderStudentIcon';

export const TOUR_END_SELECTOR =
  '#react-joyride-step-0 > div > div > div.react-joyride__tooltip > div:nth-child(2) > div > button';

export const DASHBOARD_ACTION_EDITOR_ID = 'dashboardActionEditor';
export const DASHBOARD_BAR_CHART_ID = 'dashboardBarChart';
export const DASHBOARD_USER_FILTER_ID = 'dashboardUserFilter';
export const DASHBOARD_LINE_CHART_ID = 'dashboardLineChart';
export const DASHBOARD_NO_ACTION_MESSAGE_ID = 'dashboardNoActionMessage';
export const DASHBOARD_PIE_CHART_ID = 'dashboardPieChart';
export const DASHBOARD_SPACE_FILTER_ID = 'dashboardSpaceFilter';
export const DASHBOARD_TOTAL_COUNT_ID = 'dashboardTotalCount';

export const buildPhaseTextId = (id) => `phaseText-${id}`;
export const buildSpaceDescriptionId = (id) => `spaceDescription-${id}`;
export const buildPhaseItemDescriptionId = (id) => `phaseItemDescription-${id}`;

export const FONT_SIZE_SETTING_LABEL_ID = 'fontSizeSettingLabel';
