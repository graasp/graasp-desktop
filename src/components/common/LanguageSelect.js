import React, { Component } from 'react';
import FormControl from '@material-ui/core/FormControl';
import InputLabel from '@material-ui/core/InputLabel';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import { withStyles } from '@material-ui/core/styles';
import { withTranslation } from 'react-i18next';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { langs } from '../../config/i18n';
import { getLanguage, setLanguage } from '../../actions';
import Loader from './Loader';

const styles = theme => ({
  formControl: {
    margin: theme.spacing(),
    minWidth: 120,
  },
});

export class LanguageSelect extends Component {
  static propTypes = {
    activity: PropTypes.bool.isRequired,
    i18n: PropTypes.shape({
      changeLanguage: PropTypes.func.isRequired,
    }).isRequired,
    lang: PropTypes.string.isRequired,
    t: PropTypes.func.isRequired,
    dispatchSetLanguage: PropTypes.func.isRequired,
    dispatchGetLanguage: PropTypes.func.isRequired,
    classes: PropTypes.shape({
      formControl: PropTypes.string.isRequired,
    }).isRequired,
  };

  constructor(props) {
    super(props);
    const { dispatchGetLanguage } = this.props;
    dispatchGetLanguage();
  }

  handleChangeLanguage = async ({ target }) => {
    const { i18n, dispatchSetLanguage } = this.props;
    const { value: lang } = target;
    await i18n.changeLanguage(lang);
    dispatchSetLanguage({ lang });
  };

  renderLanguageOptions = () =>
    Object.keys(langs).map(lang => (
      <MenuItem value={lang} key={lang}>
        <em>{langs[lang]}</em>
      </MenuItem>
    ));

  render() {
    const { classes, t, lang, activity } = this.props;

    if (activity) {
      return <Loader />;
    }

    return (
      <FormControl className={classes.formControl}>
        <InputLabel htmlFor="lang">{t('Language')}</InputLabel>
        <Select
          value={lang}
          onChange={this.handleChangeLanguage}
          inputProps={{
            name: 'lang',
            id: 'lang',
          }}
        >
          {this.renderLanguageOptions()}
        </Select>
      </FormControl>
    );
  }
}

const mapStateToProps = ({ User }) => ({
  lang: User.getIn(['current', 'lang']),
  activity: Boolean(User.getIn(['current', 'activity']).size),
});

const mapDispatchToProps = {
  dispatchGetLanguage: getLanguage,
  dispatchSetLanguage: setLanguage,
};

const ConnectedComponent = connect(
  mapStateToProps,
  mapDispatchToProps
)(LanguageSelect);

const StyledComponent = withStyles(styles)(ConnectedComponent);

const TranslatedComponent = withTranslation()(StyledComponent);

export default TranslatedComponent;
