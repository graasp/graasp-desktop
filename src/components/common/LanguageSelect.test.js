import React from 'react';
import { shallow } from 'enzyme';
import InputLabel from '@material-ui/core/InputLabel';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import { LanguageSelect } from './LanguageSelect';
import { langs } from '../../config/i18n';
import Loader from './Loader';

const createLanguageSelectProps = (lang, activity = false) => {
  return {
    activity,
    i18n: {
      changeLanguage: jest.fn(),
    },
    lang,
    t: text => text,
    dispatchSetLanguage: jest.fn(),
    dispatchGetLanguage: jest.fn(),
    classes: {
      formControl: '',
    },
  };
};

describe('<LanguageSelect />', () => {
  const langValues = ['en', 'fr', 'de'];

  describe.each(langValues)('<LanguageSelect /> renders', lang => {
    let wrapper;

    beforeAll(() => {
      const props = createLanguageSelectProps(lang);
      // eslint-disable-next-line react/jsx-props-no-spreading
      wrapper = shallow(<LanguageSelect {...props} />);
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders one <InputLabel /> component with text "Language"', () => {
      const inputLabel = wrapper.find(InputLabel);
      expect(inputLabel).toHaveLength(1);
      expect(inputLabel.contains('Language')).toBeTruthy();
    });

    it('renders one <Select /> component with correct language value', () => {
      const select = wrapper.find(Select);
      expect(select).toHaveLength(1);

      // check select displays all langs
      Object.keys(langs).forEach(langKey => {
        const menuItem = select.find(MenuItem).find({ value: langKey });
        expect(menuItem).toHaveLength(1);
        expect(menuItem.contains(langs[langKey])).toBeTruthy();
      });

      expect(select.prop('value')).toEqual(lang);
    });
  });

  describe.each(langValues)('<LanguageSelect /> on activity', lang => {
    it('renders <Loader />', () => {
      const props = createLanguageSelectProps(lang, true);
      // eslint-disable-next-line react/jsx-props-no-spreading
      const wrapper = shallow(<LanguageSelect {...props} />);
      expect(wrapper.find(Loader)).toHaveLength(1);
    });
  });
});
