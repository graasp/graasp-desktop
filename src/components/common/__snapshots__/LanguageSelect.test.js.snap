// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LanguageSelect /> <LanguageSelect /> renders renders correctly 1`] = `
<WithStyles(ForwardRef(FormControl))
  className=""
>
  <WithStyles(ForwardRef(InputLabel))
    htmlFor="lang"
  >
    Language
  </WithStyles(ForwardRef(InputLabel))>
  <WithStyles(ForwardRef(Select))
    inputProps={
      Object {
        "id": "lang",
        "name": "lang",
      }
    }
    onChange={[Function]}
    value="en"
  >
    <WithStyles(ForwardRef(MenuItem))
      key="en"
      value="en"
    >
      <em>
        English
      </em>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="fr"
      value="fr"
    >
      <em>
        Français
      </em>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="ja"
      value="ja"
    >
      <em>
        日本語
      </em>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="sw"
      value="sw"
    >
      <em>
        Kiswahili
      </em>
    </WithStyles(ForwardRef(MenuItem))>
  </WithStyles(ForwardRef(Select))>
</WithStyles(ForwardRef(FormControl))>
`;

exports[`<LanguageSelect /> <LanguageSelect /> renders renders correctly 2`] = `
<WithStyles(ForwardRef(FormControl))
  className=""
>
  <WithStyles(ForwardRef(InputLabel))
    htmlFor="lang"
  >
    Language
  </WithStyles(ForwardRef(InputLabel))>
  <WithStyles(ForwardRef(Select))
    inputProps={
      Object {
        "id": "lang",
        "name": "lang",
      }
    }
    onChange={[Function]}
    value="fr"
  >
    <WithStyles(ForwardRef(MenuItem))
      key="en"
      value="en"
    >
      <em>
        English
      </em>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="fr"
      value="fr"
    >
      <em>
        Français
      </em>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="ja"
      value="ja"
    >
      <em>
        日本語
      </em>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="sw"
      value="sw"
    >
      <em>
        Kiswahili
      </em>
    </WithStyles(ForwardRef(MenuItem))>
  </WithStyles(ForwardRef(Select))>
</WithStyles(ForwardRef(FormControl))>
`;

exports[`<LanguageSelect /> <LanguageSelect /> renders renders correctly 3`] = `
<WithStyles(ForwardRef(FormControl))
  className=""
>
  <WithStyles(ForwardRef(InputLabel))
    htmlFor="lang"
  >
    Language
  </WithStyles(ForwardRef(InputLabel))>
  <WithStyles(ForwardRef(Select))
    inputProps={
      Object {
        "id": "lang",
        "name": "lang",
      }
    }
    onChange={[Function]}
    value="de"
  >
    <WithStyles(ForwardRef(MenuItem))
      key="en"
      value="en"
    >
      <em>
        English
      </em>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="fr"
      value="fr"
    >
      <em>
        Français
      </em>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="ja"
      value="ja"
    >
      <em>
        日本語
      </em>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="sw"
      value="sw"
    >
      <em>
        Kiswahili
      </em>
    </WithStyles(ForwardRef(MenuItem))>
  </WithStyles(ForwardRef(Select))>
</WithStyles(ForwardRef(FormControl))>
`;
